File: java/src/main/java/com/ywh/problem/leetcode/easy/LeetCode558.java
package com.ywh.problem.leetcode.easy;

import com.ywh.model.QuadTreeNode;

/**
 * 四叉树交集
 * [树]
 *
 * @author ywh
 * @since 2019/12/4/004
 */
public class LeetCode558 {

    /**
     * Time: O(n^4), Space: O(1)
     *
     * @param quadTree1
     * @param quadTree2
     * @return
     */
    public QuadTreeNode intersect(QuadTreeNode quadTree1, QuadTreeNode quadTree2) {

        // 如果其中一个节点是叶节点，该节点值为 true，返回这个节点，否则返回另一个节点
        if (quadTree1.isLeaf) {
            return quadTree1.val? quadTree1: quadTree2;
        }
        if (quadTree2.isLeaf) {
            return quadTree2.val? quadTree2: quadTree1;
        }

        // 两个节点都不是叶节点，则继续递归
        QuadTreeNode topLeft = intersect(quadTree1.topLeft, quadTree2.topLeft);
        QuadTreeNode topRight = intersect(quadTree1.topRight, quadTree2.topRight);
        QuadTreeNode bottomLeft = intersect(quadTree1.bottomLeft, quadTree2.bottomLeft);
        QuadTreeNode bottomRight = intersect(quadTree1.bottomRight, quadTree2.bottomRight);

        // 如果四个子节点都是叶子节点，且都为 true，返回叶子节点
        if (topLeft.isLeaf && topRight.isLeaf && bottomLeft.isLeaf && bottomRight.isLeaf && topLeft.val && topRight.val && bottomLeft.val && bottomRight.val) {
            return new QuadTreeNode(true, true, null, null, null, null);
        }

        // 否则返回非叶子节点
        return new QuadTreeNode(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }
}

File: python/src/problem/leetcode/easy/leetcode_558.py
# coding=utf-8
from src.data_structure.data_structure import QuadTreeNode


class Solution:
    """
    四叉树交集
    """

    def intersect(self, quadTree1: QuadTreeNode,
                  quadTree2: QuadTreeNode) -> QuadTreeNode:
        """
        Time: O(n^4), Space: O(1)
        :param quadTree1:
        :param quadTree2:
        :return:
        """
        if quadTree1.is_leaf:
            return quadTree1 if quadTree1.val else quadTree2
        if quadTree2.is_leaf:
            return quadTree2 if quadTree2.val else quadTree1
        top_left = self.intersect(quadTree1.top_left, quadTree2.top_left)
        top_right = self.intersect(quadTree1.top_right, quadTree2.top_right)
        bottom_left = self.intersect(quadTree1.bottom_left, quadTree2.bottom_left)
        bottom_right = self.intersect(quadTree1.bottom_right, quadTree2.bottom_right)

        if top_left.is_leaf and top_right.is_leaf and bottom_left.is_leaf and\
                bottom_right.is_leaf and top_left.val and top_right.val and \
                bottom_left.val and bottom_right.val:
            return QuadTreeNode(True, True, None, None, None, None)

        return QuadTreeNode(False, False, top_left, top_right, bottom_left, bottom_right)

